### ISSUE TITLE
Improved PVSS Processing Speed

### ISSUE THREAD

**Issue by S-Hanly** [MideTechnology/endaq-python#155]
My understanding is that the PVSS calculation requires a few Python loops which makes it VERY slow, it is by far the slowest operation of all of our analysis offerings. And it is also a common one used in the cloud and by virtually all shock analysts.

Looking into using some compiled C code under-the-hood (@SamRagusa ) as well as any other performance improvements (@CFlaniganMide ) would be a noticeable improvement!

@CrepeGoat, do you have any benchmark data you can share?

**Comment by CFlaniganMide** [comment]
The calculation can definitely be sped up by multiprocessing, I've tried this with a C library I write using openMP to to the process pool allocation, and the speedup is generally a factor of how many cores your CPU has.  Another experiment showed that using AVX/SSE instructions (modern SIMD operations available on most CPUs) could produce a x4 speedup on their own.  So we have some hypothetically optimal values, but tons more work would need to be done to make it reliable.

**Comment by CFlaniganMide** [comment]
The barge shock from one of the recent webinars is probably a good real-world example: https://colab.research.google.com/drive/1cuZa5Yx55qXLhhnBMdzsJ0iiklwVi5Mq#scrollTo=dYKMz2cS6Mg_

I'd be really curious how much of a speedup we get just by adding Numba multiprocessing to the for loop that does the processing https://github.com/MideTechnology/endaq-python/blob/d2196257c96ce6e0297d01de33c65dbd1c8340ea/endaq/calc/shock.py#L202

**Comment by SamRagusa** [comment]
A few things. 

- Are you talking about the code block after the text: "Now perform the shock response spectrum calculation for those frequencies."?   Because that ran so fast that even doing a few thousand files should only take a minute or two.

- Unfortunately I think you have some misconceptions about how Numba's JIT compilation works.  Almost none of the lines of code you attached would be able to successfully compile without change (and it's an all-or-none situation).  A (pretty large) [subset of pure python](https://numba.pydata.org/numba-doc/dev/reference/pysupported.html) is supported, some standard library functionality is supported, and a [large subset of the NumPy package](https://numba.pydata.org/numba-doc/dev/reference/numpysupported.html) is supported.    _That is it._  Even if you did want to 'lower' the SciPy functions so that they could be called from within the nopython mode of Numba's compiled code, it still wouldn't be compiling those functions you've lowered with the rest of the code, and thus I don't think it'd see big speed ups on this particular example (since the overhead of calling the SciPy functions is likely negligible).   

The way to utilize Numba to get those crazy several order of magnitude speed improvements I like to boast about is not to have it compile 
```Python
def rms(x):
    return np.sqrt(np.mean(np.square(x))))
```

but instead to have it compile something like this (not that code specifically, but code written like that):

```Python
def rms(x):
    total = 0
    for num in x:
        total += num ** 2
    total  /= len(x)
return np.sqrt(total)
```

So basically writing out what those SciPy functions do in simple explicit ways would be the way forward with Numba.

Contextual awareness of the compiler is absolutely key.

**Comment by SamRagusa** [comment]
You're misunderstanding what that flag does, if `parrallel=True` is set and you're using `numba.jit` for JIT compilation,  then the code **will fail** if it needs to fall back from 'nopython' mode to 'object mode' (aka if it would fail if the `nopython=True` parameter is set, or if you just use `numba.njit` which I prefer since it's more obvious what it's doing).  You can not just put arbitrary code in parallel with Numba.  In fact it's a smaller subset of what works in regular nopython mode that.

Source: https://numba.pydata.org/numba-doc/latest/user/jit.html#compilation-options

Also anecdotally, I've spent many, many hours trying to squeeze an extra bit of performance out of using these flags and never got any performance worth the overhead added (code still runs on multiple cores even with `parallel=False`).

**Comment by shanlyMIDE** [comment]
Yeah the problem doesn't really exist for a single PVSS calculation on a ~1 second long dataset. But it compounds when:
- Processing many files like in `endaq.batch`
- Computing the PVSS on a long time series, say 10 minutes, even 10 hours long - it can take literally hours

On the second point, the best approach is actually to force the PVSS to find maybe the "worst" ~5 events in the recording and compute the PVSS just for those events, not the entire time series. "Worst" becomes a question as to how you define, RMS per 0.5 seconds window was recommended by Hutchinson shock analysis experts. Note that I made a issue request for this #156

**Comment by CFlaniganMide** [comment]
> You're misunderstanding what that flag does, if `parrallel=True` is set and you're using `numba.jit` for JIT compilation, then the code **will fail** if it needs to fall back from 'nopython' mode to 'object mode' (aka if it would fail if the `nopython=True` parameter is set, or if you just use `numba.njit` which I prefer since it's more obvious what it's doing). You can not just put arbitrary code in parallel with Numba. In fact it's a smaller subset of what works in regular nopython mode that.
> 
> Source: https://numba.pydata.org/numba-doc/latest/user/jit.html#compilation-options
> 
> Also anecdotally, I've spent many, many hours trying to squeeze an extra bit of performance out of using these flags and never got any performance worth the overhead added (code still runs on multiple cores even with `parallel=False`).

Gotcha, I missed that.  A bigger lift for not much return.

**Comment by pscheidler** [comment]
This issue has a lower priority because the focus will be on separating out shock events and then running PVSS on just the shock events.
